File: algorithms/warmup/grading/grading-python/tests/gradingStudents.py
# O problema em questão no HackerRank, 
# com o título "Grading Students" (Arredondamento de Notas), 
# pede que você implemente uma função que realize o arredondamento 
# de notas de estudantes de acordo com as seguintes regras:

# Se a diferença entre a nota e o próximo múltiplo 
# de 5 for menor que 3, arredonde a nota para o próximo 
# múltiplo de 5.
# Se a nota for menor que 38, não faça nenhum 
# arredondamento, pois a diferença entre a nota e o 
# próximo múltiplo de 5 seria maior ou igual a 3.
# A entrada do problema consiste em uma lista de notas, 
# e a saída esperada é uma nova lista de notas após o arredondamento.

# def gradingStudents(grades):
    
#     # Itera sobre a lista de notas
#     for i in range(len(grades)):
        
#         # Verifica se a nota é maior ou igual a 38 para ser arredondada
#         if grades[i] >= 38:
            
#             # Calcula o resto da divisão da nota por 5
#             lembrar = grades[i] % 5
            
#             # Se o resto for maior ou igual a 3, arredonda a nota para o próximo múltiplo de 5
#             if lembrar >= 3:
                
#                 # Adiciona a diferença para arredondar a nota
#                 grades[i] += 5 - lembrar
    
#     # Retorna a lista de notas modificada
#     return grades


# def gradingStudents(grades):
    
#     # Lista para armazenar as notas corrigidas
#     arr = []
    
#     # Itera sobre as notas na lista original
#     for i in grades:
        
#         # Verifica se a nota é menor que 38
#         if i < 38:
            
#             # Se for menor que 38, 
#             # a nota permanece inalterada e é adicionada à lista corrigida
#             arr.append(i)
            
#         # Verifica se a diferença entre o próximo múltiplo de 5 e a nota é menor que 3
#         elif (((i // 5) + 1) * 5  - i) < 3:
            
#             # Se a diferença for menor que 3, 
#             # arredonda a nota para o próximo múltiplo de 5
#             arr.append(((i // 5) + 1) * 5)
            
#         else:
            
#             # Se não atender às condições anteriores, 
#             # a nota permanece inalterada e é adicionada à lista corrigida
#             arr.append(i)
    
#     # Retorna a lista de notas corrigidas
#     return arr


def gradingStudents(grades):
    
    # Lista para armazenar as notas arredondadas
    notas_arredondadas = []
    
    # Itera sobre as notas na lista original
    for grade in grades:
        
        # Verifica se a nota é maior ou igual a 38 para ser arredondada
        if grade >= 38:
            
            # Calcula a diferença até o próximo múltiplo de 5
            diferenca = 5 - grade % 5
            
            # Se a diferença for menor que 3, arredonda a nota
            if diferenca < 3:
                
                # Adiciona a diferença para arredondar a nota
                grade += diferenca
                
        # Adiciona a nota (arredondada ou não) à lista de notas arredondadas
        notas_arredondadas.append(grade)
        
    # Retorna a lista de notas arredondadas
    return notas_arredondadas
    
    

File: algorithms/warmup/grading/grading-javscript/js/gradingStudents.js
// Função para arredondar as notas conforme as regras especificadas
// export function gradingStudents(grades) {
    
//     // Loop sobre todas as notas no array
//     for (let index = 0; index < grades.length; index++) {

//         // Calcula o próximo múltiplo de 5 para a nota atual
//         const element = Math.ceil(grades[index] / 5) * 5;

//         // Verifica se a nota é maior ou igual a 38 e se a diferença entre o próximo múltiplo de 5 e a nota é menor que 3
//         if (grades[index] >= 38 && element - grades[index] < 3) {
            
//             // Se as condições são atendidas, substitui a nota pela próxima múltiplo de 5
//             grades[index] = element;
//         }
//     }

//     // Retorna o array de notas processado
//     return grades;
// }


export function gradingStudents(grades) {

    // Utiliza o método map para aplicar a lógica de arredondamento a cada nota na lista
    return grades.map((grade) => {

        // Calcula a diferença até o próximo múltiplo de 5
        let novaGrande = Math.ceil(grade / 5) * 5 - grade;

        // Retorna a nova nota arredondada ou a nota original, dependendo das condições
        return (novaGrande < 3 && grade >= 38) ? Math.ceil(grade / 5) * 5 : grade
    });
}